File: Question8/Question8.py
def solution(n, m, costs):
    cost=[]
    for i in range(n+1):
            cost.append([0]*(m+1))
    for i in range(1, n + 1):
        prevCost= float('-inf')
        for j in range(1, m):
            prevCost = max(prevCost,cost[i - 1][j - 1]-costs[j - 1])
            cost[i][j] = max(cost[i][j - 1],costs[j] + prevCost)
    sumOfCost=cost[n][m-1]
    return sumOfCost
    
def main():
    n, m=map(int,input().split())
    costs = list(map(int, input().split()))
    print(solution(n, m, costs))


if __name__ == "__main__":
    main()

File: Question8/Question8.cpp
#include <bits/stdc++.h>
using namespace std;
void solution(int n, int m, int *costs) {
    int prevCost,cost[n+1][m];
    memset(cost, 0, sizeof cost); 
    for (int i = 1; i <= n; i++) {
        prevCost = INT_MIN;
        for (int j = 1; j <m; j++) {
            prevCost = max(prevCost,*(*(cost + i-1) + j-1) - *(costs+j - 1));
            *(*(cost + i) + j)= max(*(*(cost + i) + j-1), *(costs+j)+ prevCost);
            }}
    printf("%d\n",*(*(cost + n) + m-1));
    return;
}
int main() {
    int n,m;
    cin>>n>>m;
    int costs[m];
    for(int i=0;i<m;i++){
        cin>>costs[i];
    }
	solution(n,m,costs);
return 0;
}